proto/build.rs

```diff
@@ -11,131 +11,151 @@
+
+    pub fn compile<P: AsRef<Path>>(desc_path: P) {
+        desc_to_module(desc_path.as_ref());
     }
-}
 
-/// Descriptor file to module file using rust-protobuf.
-fn desc_to_module_rust_protobuf<G, W>(
-    descriptor: &Path,
-    output_dir: &Path,
-    mut generate_files: G,
-    out: W,
-) where
-    G: FnMut(&[FileDescriptorProto], &[String]) -> Vec<GenResult>,
-    W: Write,
-{
-    let mut f = File::open(descriptor).unwrap();
-    let proto_set: FileDescriptorSet = protobuf::parse_from_reader(&mut f).unwrap();
-
-    let files: Vec<_> = proto_set
-        .get_file()
-        .into_iter()
-        .map(|f| f.get_name().to_owned())
-        .collect();
-
-    // All files need to be generated in our case.
-    let results = generate_files(proto_set.get_file(), &files);
-    write_files(
-        results.into_iter().map(|res| (res.name, res.content)),
-        output_dir,
-        out,
-    );
+    /// Descriptor file to module file using Prost.
+    fn desc_to_module(descriptor: &Path) {
+        let buf = fs::read(descriptor).unwrap();
+        let proto_set: ProstFileDescriptorSet = Message::decode(buf).unwrap();
+        let files: Vec<_> = proto_set
+            .file
+            .into_iter()
+            .map(|f| {
+                let mut path = PathBuf::new();
+                path.push("proto");
+                path.push(f.name.unwrap());
+                path.to_str().unwrap().to_owned()
+            })
+            .collect();
+
+        let packages = protobuf_build::generate_prost_files(&files, &env::var("OUT_DIR").unwrap());
+        assert!(!packages.is_empty());
+        protobuf_build::generate_wrappers(
+            &packages
+                .iter()
+                .map(|m| format!("{}/{}.rs", env::var("OUT_DIR").unwrap(), m))
+                .collect::<Vec<_>>(),
+            &env::var("OUT_DIR").unwrap(),
+            protobuf_build::GenOpt::all(),
+        );
+    }
 }
 
-/// Descriptor file to module file using Prost.
-fn desc_to_module_prost(descriptor: &Path) {
-    let buf = fs::read(descriptor).unwrap();
-    let proto_set: ProstFileDescriptorSet = Message::decode(buf).unwrap();
-    let files: Vec<_> = proto_set
-        .file
-        .into_iter()
-        .map(|f| {
-            let mut path = PathBuf::new();
-            path.push("proto");
-            path.push(f.name.unwrap());
-            path.to_str().unwrap().to_owned()
-        })
-        .collect();
-
-    let packages = protobuf_build::generate_prost_files(&files, &env::var("OUT_DIR").unwrap());
-    assert!(!packages.is_empty());
-    protobuf_build::generate_wrappers(
-        &packages
-            .iter()
-            .map(|m| format!("{}/{}.rs", env::var("OUT_DIR").unwrap(), m))
-            .collect::<Vec<_>>(),
-        &env::var("OUT_DIR").unwrap(),
-        protobuf_build::GenOpt::all(),
-    );
+#[cfg(not(feature = "prost-codec"))]
+mod prost {
+    use std::path::Path;
+
+    pub fn compile<P: AsRef<Path>>(_desc_path: P) {}
```

Comment:
Why is this needed?

Reply:
I think it's related to https://github.com/pingcap/grpc-rs/pull/343#discussion_r304208236

Reply:
Yeah, it gets called from `main` so you need the stub function here. Alternative would be to `cfg` the call in main, I don't mind either version tbh

Reply:
Ok